metaflow/plugins/kfp/kfp.py

```diff
@@ -1,316 +1,419 @@
-                                                            code_url=code_url)], shell=True)
-
-    print("\n----------RUNNING: KFP Installation---------------")
-    subprocess.call(["pip3 install kfp"], shell=True)  # TODO: Remove this once KFP is added to dependencies
-
-    print("\n----------RUNNING: METAFLOW INSTALLATION----------")
-    subprocess.call(["pip3 install --user --upgrade {modified_metaflow_git_url}".format(
-        modified_metaflow_git_url=MODIFIED_METAFLOW_URL)], shell=True)
-
-    print("\n----------RUNNING: MAIN STEP COMMAND--------------")
-
-    S3_BUCKET = os.getenv("S3_BUCKET")
-    S3_AWS_ARN = os.getenv("S3_AWS_ARN")
-    S3_AWS_REGION = os.getenv("S3_AWS_REGION")
-
-    define_s3_env_vars = 'export METAFLOW_DATASTORE_SYSROOT_S3="{}" && export METAFLOW_AWS_ARN="{}" ' \
-                         '&& export METAFLOW_AWS_S3_REGION="{}"'.format(S3_BUCKET, S3_AWS_ARN, S3_AWS_REGION)
-    define_username = 'export USERNAME="kfp-user"' # TODO: Map username to KFP specific user/profile/namespace
-    python_cmd = python_cmd_template.format(ds_root=S3_BUCKET, run_id=kfp_run_id)
-
-    final_run_cmd = "{define_username} && {define_s3_env_vars} && {python_cmd}".format(define_username=define_username,
-                                                                                       define_s3_env_vars=define_s3_env_vars,
-                                                                                       python_cmd=python_cmd)
-
-    print("RUNNING COMMAND: ", final_run_cmd)
-    proc = subprocess.run(final_run_cmd, shell=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE, universal_newlines=True)
-    proc_output = proc.stdout
-    proc_error = proc.stderr
-
-    if len(proc_error) > 1:
-        print("_______________STDERR:_____________________________")
-        print(proc_error)
-
-    if len(proc_output) > 1:
-        print("_______________STDOUT:_____________________________")
-        print(proc_output)
-
-    # TODO: Metadata needed for client API to run needs to be persisted outside before return
-
-    print("_______________ Done _________________________________")
-    # END is the final step
-    if step_name.lower() == 'end':
-        print("_______________ FLOW RUN COMPLETE ________________")
-
-
-def start_op_func(start_command_template: str, code_url: str, kfp_run_id: str):
-    """
-    Function used to create a KFP container op corresponding to the 'start' step of the flow.
-    This function also defines the execution of an init step which is needed before the 'start' step
-    executes in order to persist parameters of the flow (i.e., class level variables).
-    """
-    import subprocess
-    import os
+import kfp
 
-    def execute(cmd):
+from .constants import DEFAULT_KFP_YAML_OUTPUT_PATH
+from ... import R
+from ...graph import DAGNode
+
+
+class KubeflowPipelines(object):
+    def __init__(
+        self,
+        name,
+        graph,
```

Comment:
Can we add type annotations here? Especially for parameters like `graph` and `flow`, it would help to go to the actual object.

Reply:
I've been avoiding adding type annotations because the Metaflow codebase also supports Python 2.7

Reply:
Makes sense.